File: app/lib/models/recipe.dart
class Recipe {
  final String chefkochId;
  final String title;
  final String? imageUrl;
  final int prepTime;
  final int cookTime;
  final String? difficulty;
  final double rating;
  final String url;
  final int servings;
  final List<Map<String, dynamic>> ingredients;
  bool isFavorite;
  int? savedId;

  Recipe({
    required this.chefkochId,
    required this.title,
    this.imageUrl,
    this.prepTime = 0,
    this.cookTime = 0,
    this.difficulty,
    this.rating = 0.0,
    required this.url,
    this.servings = 4,
    this.ingredients = const [],
    this.isFavorite = false,
    this.savedId,
  });

  factory Recipe.fromJson(Map<String, dynamic> j) => Recipe(
        chefkochId: j['chefkoch_id'] as String? ?? j['id']?.toString() ?? '',
        title: j['title'] as String? ?? '',
        imageUrl: j['image_url'] as String?,
        prepTime: j['prep_time'] as int? ?? 0,
        cookTime: j['cook_time'] as int? ?? 0,
        difficulty: j['difficulty'] as String?,
        rating: (j['rating'] as num?)?.toDouble() ?? 0.0,
        url: j['url'] as String? ?? '',
        servings: j['servings'] as int? ?? 4,
        ingredients: (j['ingredients'] as List<dynamic>?)
                ?.map((e) => Map<String, dynamic>.from(e as Map))
                .toList() ??
            [],
        isFavorite: j['is_favorite'] as bool? ?? false,
        savedId: j['id'] as int?,
      );

  String get timeLabel {
    final total = prepTime + cookTime;
    if (total == 0) return '';
    if (total < 60) return '${total} Min.';
    return '${total ~/ 60}h ${total % 60}min';
  }
}

File: api/schemas/recipe.py
from datetime import datetime
from typing import Optional
from pydantic import ConfigDict, BaseModel, Field


class RecipeSearchResult(BaseModel):
    chefkoch_id: str = Field(..., min_length=1, max_length=200)
    title: str = Field(..., min_length=1, max_length=200)
    image_url: Optional[str] = Field(None, max_length=500)
    prep_time: int = Field(0, ge=0)
    cook_time: int = Field(0, ge=0)
    difficulty: Optional[str] = Field(None, max_length=50)
    rating: float = Field(0.0, ge=0.0, le=5.0)
    url: str = Field(..., max_length=500)


class RecipeFull(RecipeSearchResult):
    servings: int = Field(4, ge=1)
    ingredients: list[dict] = []
    instructions: Optional[str] = Field(None, max_length=10000)


class SavedRecipeCreate(BaseModel):
    chefkoch_id: str = Field(..., min_length=1, max_length=200)
    title: str = Field(..., min_length=1, max_length=200)
    image_url: Optional[str] = Field(None, max_length=500)
    servings: int = Field(4, ge=1)
    prep_time: int = Field(0, ge=0)
    cook_time: int = Field(0, ge=0)
    difficulty: Optional[str] = Field(None, max_length=50)
    ingredients_json: Optional[str] = Field(None, max_length=20000)
    source_url: Optional[str] = Field(None, max_length=500)


class SavedRecipeOut(SavedRecipeCreate):
    id: int
    user_key: str
    is_favorite: bool
    created_at: datetime

    model_config = ConfigDict(from_attributes=True)


class ToShoppingRequest(BaseModel):
    servings: int = Field(4, ge=1, le=100)  # Für Portionskalkulation


class IngredientItem(BaseModel):
    name: str = Field(..., min_length=1, max_length=200)
    amount: Optional[str] = None
    unit: Optional[str] = None


class SelectedIngredientsRequest(BaseModel):
    ingredients: list[IngredientItem] = Field(..., min_length=1)
